tensorpipe/channel/cuda_basic/context.cc

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/channel/cuda_basic/context.h>
+
+#include <algorithm>
+#include <list>
+
+#include <tensorpipe/channel/cuda_basic/channel.h>
+#include <tensorpipe/channel/cuda_basic/context_impl.h>
+#include <tensorpipe/channel/error.h>
+#include <tensorpipe/channel/helpers.h>
+#include <tensorpipe/common/callback.h>
+#include <tensorpipe/common/cuda_loop.h>
+#include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error.h>
+#include <tensorpipe/common/error_macros.h>
+
+namespace tensorpipe {
+namespace channel {
+namespace cuda_basic {
+
+class Context::Impl : public Context::PrivateIface,
+                      public std::enable_shared_from_this<Context::Impl> {
+ public:
+  Impl(std::shared_ptr<CpuContext>);
+
+  const std::string& domainDescriptor() const;
+
+  std::shared_ptr<channel::CudaChannel> createChannel(
+      std::shared_ptr<transport::Connection>,
+      Endpoint);
+
+  void setId(std::string id);
+
+  ClosingEmitter& getClosingEmitter() override;
+
+  void close();
+
+  void join();
+
+  ~Impl() override = default;
+
+ private:
+  std::string domainDescriptor_;
+  std::atomic<bool> closed_{false};
+  std::atomic<bool> joined_{false};
+  ClosingEmitter closingEmitter_;
+
+  std::shared_ptr<CpuContext> cpuContext_;
+  std::shared_ptr<CudaLoop> cudaLoop_;
```

Comment:
Is there a reason to wrap this in a shared_ptr?

Reply:
I guess it's so that channels can keep a copy of the shared_ptr and keep the CudaLoop alive? In principle there shouldn't be a need for it: channels already hold a shared_ptr to the Context::Impl and thus keep that alive.